convert_hf_to_gguf.py

```diff
@@ -1514,6 +1514,35 @@ def set_gguf_parameters(self):
         if self.hparams.get("vocab_size", 32000) == 49152:
             self.gguf_writer.add_add_bos_token(False)
 
+        if rope_scaling := self.find_hparam(["rope_scaling"], optional=True):
+            if rope_scaling.get("rope_type", '').lower() == "llama3":
+                base = hparams.get("rope_theta", 10000.0)
+                dim = int((hparams["hidden_size"] // hparams["num_attention_heads"]) * hparams.get("partial_rotary_embeddings", 1.0))
+                inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2, dtype=torch.float32) / dim))
+
+                factor = rope_scaling.get("factor", 8.0)
+                low_freq_factor = rope_scaling.get("low_freq_factor", 1.0)
+                high_freq_factor = rope_scaling.get("high_freq_factor", 4.0)
+                old_context_len = hparams.get("original_max_position_embeddings", 8192)
+
+                low_freq_wavelen = old_context_len / low_freq_factor
+                high_freq_wavelen = old_context_len / high_freq_factor
+
+                rope_freqs = []
+                for freq in inv_freq:
+                    wavelen = 2 * math.pi / freq
+                    if wavelen < high_freq_wavelen:
+                        rope_freqs.append(freq)
+                    elif wavelen > low_freq_wavelen:
+                        rope_freqs.append(freq / factor)
+                    else:
+                        assert low_freq_wavelen != high_freq_wavelen
+                        smooth = (old_context_len / wavelen - low_freq_factor) / (high_freq_factor - low_freq_factor)
+                        rope_freqs.append((1 - smooth) * freq / factor + smooth * freq)
+
+                self.gguf_writer.add_rope_scaling_attn_factors(1.0)
+                self.gguf_writer.add_tensor(gguf.TENSOR_NAMES[gguf.MODEL_TENSOR.ROPE_FREQS] + ".weight", 1.0 / np.array(rope_freqs))
```

Comment:
Unlike the transformers or pytorch reference code, I believe we need to invert `rope_freqs` here, since they applied later in an inverted way in the rope kernels

Reply:
The `freq_factors` in ggml.c are divided into `theta`, which is $p b^{-2 i/d}$ https://github.com/ggerganov/llama.cpp/blob/68504f0970db5a3602d176953690f503059906b1/ggml/src/ggml-metal.metal#L1567-L1571 and then taken `cos` and `sin`. But the tensor here you created is $f / b^{-2 i/d}$. Perhaps the factors should be built with `factors.append(1)`, `factors.append(factor)`, `factors.append(1/((1-s)*factor+s))` depending on the wavelen?